keps/0024-parameter-enhancement.md

```diff
@@ -0,0 +1,146 @@
+---
+kep-number: 24
+title: Enabling and Disabling features in KUDO Operators
+authors:
+  - "@nfnt"
+owners:
+  - "@nfnt"
+creation-date: 2020-02-21
+last-updated: 2020-02-21
+status: provisional
+---
+
+# Enhanced Operator Parameters
+
+## Table of Contents
+
+- [Enhanced Operator Parameters](#enhanced-operator-parameters)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [Support Optional Parameter Types](#support-optional-parameter-types)
+    - [Provide Parameter Validation](#provide-parameter-validation)
+    - [Alternative Considered: Convert YAML in string to dictionaries](#alternative-considered-convert-yaml-in-string-to-dictionaries)
+    - [Alternative Considered: Allow Arbitrary Parameter Values](#alternative-considered-allow-arbitrary-parameter-values)
+  - [Implementation History](#implementation-history)
+
+## Summary
+
+This KEP aims to improve the use-cases that can be covered by parameters. By supporting additional types, templates can access lists and dictionaries. Furthermore, validation can be implemented for known types.
+
+## Motivation
+
+Currently only string parameters are supported by KUDO. This limits the possible use-cases of parameters. Various use cases benefit from having parameters that are lists or dictionaries. While some of this functionality can be implemented in Go templates using [Sprig functions][1], this is unintuitive.
+Having parameter types creates the possibility for parameter validation.
+
+### Goals
+
+- Provide operator developers with more versatile parameters
+- Allow the definition of validation functions for parameters
+
+### Non-Goals
+
+- Change how parameters are defined
+
+## Proposal
+
+### Support Optional Parameter Types
+
+By adding an optional `type` field to the list of parameters, parameter types can be specified. If this field isn't present, it defaults to the `string` type. Other types like `list`, `dict` or `integer` are possible. KUDO's renderer uses the type information to convert the string value to the respective type. For example, for the `dict` type, the input value would be YAML that is unmarshalled to a `map[string]interface{}`. Go's template engine allows to access and iterate list and dictionary items.
+
+Consider a cluster that has nodes in multiple regions. We want to ensure that a nginx deployment has 2 replicas in one region and 1 replica in the other one. Furthermore they should support both HTTP as well as HTTPS, hence need a list of port definitions. These parameters are described by `list` and `dict` types:
+
+```yaml
+parameters:
+  - name: ports
+    description: Ports
+    type: list
+    default: [ 80, 443 ]
+  - name: topology
+    description: Node topology
+    type: dict
+    default: |
```

Comment:
Hm, so a `list` is a native YAML list, but `dict` is a string that happens to be formatted like a YAML list?
This seems inconsistent.

Reply:
Both are formatted as YAML in `params.yaml`. And both could be covered by a `yaml` type. The reason to distinguish between these types here is that additional validation functions could be different for the types. But without any validation they're both equivalent because they're both provided as YAML.

Reply:
Not sure what you mean by:

> Both are formatted as YAML

Clearly the default value for `topology` is a string literal, [according to the YAML spec for the pipe symbol `|`](https://yaml.org/spec/1.2/spec.html#id2795688), while the default value for `ports` is a list of integers. The fact that the string literal happens to be in YAML format is irrelevant (I don't think we want multiple layers of YAML parsing?)

Reply:
Now I see what you meant :)
I'll remove the pipe symbol.